tpls/desul/include/desul/atomics/Compare_Exchange_OpenACC.hpp

```diff
@@ -0,0 +1,269 @@
+  T return_val;
+#pragma acc atomic capture
+  {
+    return_val = *dest;
+    *dest = value;
+  }
+  return return_val;
+}
+
+#endif
+
+
+#ifdef KOKKOS_COMPILER_NVHPC
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<!std::is_arithmetic<T>::value || ((sizeof(T) != 4) && (sizeof(T) != 8)), T> 
+	device_atomic_compare_exchange(T* dest, T compare, T value, MemoryOrder, MemoryScope scope) {
+  T current_val = *dest;
+  if (acc_on_device(acc_device_not_host)) {
+    printf("DESUL error in device_atomic_compare_exchange(): Not supported atomic "
+                  "operation in the OpenACC backend\n");
+  }
+  // Acquire a lock for the address
+  //while (!lock_address_openacc((void*)dest, scope)) {
+  //}
+  //device_atomic_thread_fence(MemoryOrderAcquire(), scope);
+  if (current_val == compare) {
+    *dest = value;
+    //device_atomic_thread_fence(MemoryOrderRelease(), scope);
+  }
+  //unlock_address_openacc((void*)dest, scope);
+  return current_val;
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_integral<T>::value && (sizeof(T) == 4), T> device_atomic_compare_exchange(T* const dest, T compare, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(unsigned int) == 4,
+                "this function assumes an unsigned int is 32-bit");
+  unsigned int return_val = atomicCAS(reinterpret_cast<unsigned int*>(dest),
+                                      reinterpret_cast<unsigned int&>(compare),
+                                      reinterpret_cast<unsigned int&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_integral<T>::value && (sizeof(T) == 8), T> device_atomic_compare_exchange(T* const dest, T compare, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(unsigned long long int) == 8,
+                "this function assumes an unsigned long long int is 64-bit");
+  unsigned long long int return_val = atomicCAS(reinterpret_cast<unsigned long long int*>(dest),
+                                      reinterpret_cast<unsigned long long int&>(compare),
+                                      reinterpret_cast<unsigned long long int&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_floating_point<T>::value && (sizeof(T) == 4), T> device_atomic_compare_exchange(T* const dest, T compare, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(float) == 4,
+                "this function assumes an float is 32-bit");
+  float return_val = atomicCAS(reinterpret_cast<float*>(dest),
+                                      reinterpret_cast<float&>(compare),
+                                      reinterpret_cast<float&>(value));
+  return reinterpret_cast<T&>(return_val);
```

Comment:
Same question here, is there a good reason to treat floating point separately.

Reply:
If a single implementation is used for both integer and floating types, the NVHPC compiler fails in the integer-to-float casting code (return reinterpret_cast<T&>(return_val);) with the following error:
"parse invalid cast opcode for cast from 'i32' to 'float'"